File: poolvr/physics/collisions.py
"""
This module implements the ball-ball collision model described in: ::

  NUMERICAL SIMULATIONS OF THE FRICTIONAL COLLISIONS
  OF SOLID BALLS ON A ROUGH SURFACE
  S. Mathavan,  M. R. Jackson,  R. M. Parkin
  DOI: 10.1007/s12283-014-0158-y
  International Sports Engineering Association
  2014

"""
from logging import getLogger
_logger = getLogger(__name__)
from math import sqrt
import numpy as np
from numpy import dot, cross


INCH2METER = 0.0254
INF = float('inf')
_k = np.array([0, 1, 0], dtype=np.float64)


def collide_balls(r_c,
                  r_i, v_i, omega_i,
                  r_j, v_j, omega_j,
                  e=0.89,
                  mu_s=0.21,
                  mu_b=0.05,
                  M=0.1406,
                  R=0.02625,
                  g=9.81,
                  deltaP=None,
                  return_all=False):
    r_ij = r_j - r_i
    r_ij_mag_sqrd = dot(r_ij, r_ij)
    # D = 2*R
    #assert  abs(r_ij_mag_sqrd - D**2) / D**2  <  1e-4, "abs(r_ij_mag_sqrd - D**2) / D**2 = %s" % (abs(r_ij_mag_sqrd - D**2) / D**2)
    r_ij_mag = sqrt(r_ij_mag_sqrd)
    z_loc = _k
    y_loc = r_ij / r_ij_mag
    x_loc = cross(y_loc, z_loc)
    G = np.vstack((x_loc, y_loc, z_loc))
    v_i = dot(G, v_i)
    v_j = dot(G, v_j)
    omega_i = dot(G, omega_i)
    omega_j = dot(G, omega_j)
    r_ic = np.array([0.0,  R, 0.0])
    r_jc = np.array([0.0, -R, 0.0])
    u_iR = v_i + R * cross(_k, omega_i)
    u_iR_mag = sqrt(dot(u_iR, u_iR))
    u_jR = v_j + R * cross(_k, omega_j)
    u_jR_mag = sqrt(dot(u_jR, u_jR))
    u_iC = v_i - cross(r_ic, omega_i)
    u_jC = v_j - cross(r_jc, omega_j)
    u_ijC = u_iC - u_jC
    u_ijC_xz = u_ijC[::2]
    u_ijC_xz_mag = sqrt(dot(u_ijC_xz, u_ijC_xz))
    v_ij = v_j - v_i
    if deltaP is None:
        deltaP = 0.5 * (1 + e) * M * abs(v_ij[1]) / 1000
    W_f = INF
    W_c = None
    W = 0
    niters = 0
    if return_all:
        v_is = [v_i]
        v_js = [v_j]
        omega_is = [omega_i]
        omega_js = [omega_j]
    while v_ij[1] < 0 or W < W_f:
        if u_ijC_xz_mag < 1e-16:
            deltaP_1 = deltaP_2 = 0
            deltaP_ix = deltaP_iy = deltaP_jx = deltaP_jy = 0
        else:
            deltaP_1 = -mu_b * deltaP * u_ijC[0] / u_ijC_xz_mag
            if abs(u_ijC[2]) < 1e-16:
                deltaP_2 = 0
                deltaP_ix = deltaP_iy = deltaP_jx = deltaP_jy = 0
            else:
                deltaP_2 = -mu_b * deltaP * u_ijC[2] / u_ijC_xz_mag
                if deltaP_2 > 0:
                    deltaP_ix = deltaP_iy = 0
                    if u_jR_mag == 0:
                        deltaP_jx = deltaP_jy = 0
                    else:
                        deltaP_jx = -mu_s * (u_jR[0] / u_jR_mag) * deltaP_2
                        deltaP_jy = -mu_s * (u_jR[1] / u_jR_mag) * deltaP_2
                else:
                    deltaP_jx = deltaP_jy = 0
                    if u_iR_mag == 0:
                        deltaP_ix = deltaP_iy = 0
                    else:
                        deltaP_ix = mu_s * (u_iR[0] / u_iR_mag) * deltaP_2
                        deltaP_iy = mu_s * (u_iR[1] / u_iR_mag) * deltaP_2
        # velocity changes:
        deltaV_ix = (deltaP_1  + deltaP_ix) / M
        deltaV_iy = (-deltaP   + deltaP_iy) / M
        deltaV_jx = (-deltaP_1 + deltaP_jx) / M
        deltaV_jy = (deltaP    + deltaP_jy) / M
        # angular velocity changes (due to frictional forces):
        deltaOm_i = 5/(2*M*R) * np.array([( deltaP_2 + deltaP_iy),
                                          (-deltaP_ix),
                                          (-deltaP_1)])
        deltaOm_j = 5/(2*M*R) * np.array([( deltaP_2 + deltaP_jy),
                                          (-deltaP_jx),
                                          (-deltaP_1)])
        # update velocities:
        v_i0 = v_i
        v_i = v_i0 + np.array([deltaV_ix, deltaV_iy, 0])
        v_j0 = v_j
        v_j = v_j0 + np.array([deltaV_jx, deltaV_jy, 0])
        # update angular velocities:
        omega_i0 = omega_i
        omega_i = omega_i0 + deltaOm_i
        omega_j0 = omega_j
        omega_j = omega_j0 + deltaOm_j
        # update ball-table slips:
        u_iR = v_i + R * cross(_k, omega_i)
        u_jR = v_j + R * cross(_k, omega_j)
        u_iR_mag = sqrt(dot(u_iR, u_iR))
        u_jR_mag = sqrt(dot(u_jR, u_jR))
        # update ball-ball slip:
        u_iC = v_i - cross(r_ic, omega_i)
        u_jC = v_j - cross(r_jc, omega_j)
        u_ijC = u_iC - u_jC
        u_ijC_xz = u_ijC[::2]
        u_ijC_xz_mag = sqrt(dot(u_ijC_xz, u_ijC_xz))
        # increment work:
        v_ijy0 = v_ij[1]
        v_ij = v_j - v_i
        deltaW = 0.5 * deltaP * abs(v_ijy0 + v_ij[1])
        W += deltaW
        niters += 1
        if return_all:
            v_is.append(v_i)
            v_js.append(v_j)
            omega_is.append(omega_i)
            omega_js.append(omega_j)
        if W_c is None and v_ij[1] > 0:
            W_c = W
            W_f = (1 + e**2) * W_c
            # niters_c = niters
            # _logger.debug('''
            # END OF COMPRESSION PHASE
            # W_c = %s
            # W_f = %s
            # niters_c = %s
            # ''', W_c, W_f, niters_c)
    # _logger.debug('''
    # END OF RESTITUTION PHASE
    # niters_r = %s
    # ''', niters - niters_c)
    if return_all:
        v_is = np.array(v_is)
        v_js = np.array(v_js)
        omega_is = np.array(omega_is)
        omega_js = np.array(omega_js)
        for i in range(len(v_is)):
            dot(G.T, v_is[i], out=v_is[i])
            dot(G.T, v_js[i], out=v_js[i])
            dot(G.T, omega_is[i], out=omega_is[i])
            dot(G.T, omega_js[i], out=omega_js[i])
        return v_is, omega_is, v_js, omega_js
    return dot(G.T, v_i), dot(G.T, omega_i), dot(G.T, v_j), dot(G.T, omega_j)

File: poolvr/physics/collisions.f90
MODULE collisions
  USE iso_c_binding
  IMPLICIT NONE
  double precision, parameter :: INF = atan(-1.d0)
  double precision, parameter :: R = 0.02625
  double precision, parameter :: M = 0.1406
  double precision, parameter :: mu_s = 0.21
  double precision, parameter :: mu_b = 0.05
  double precision, parameter :: e = 0.89
CONTAINS

  SUBROUTINE collide_balls (deltaP, R, r_i, v_i, omega_i, r_j, v_j, omega_j, v_i1, omega_i1, v_j1, omega_j1) BIND(C)
    USE, INTRINSIC :: ISO_C_BINDING
    implicit none
    double precision, VALUE, intent(in) :: deltaP, R
    double precision, dimension(3), intent(in) :: r_i, r_j
    double precision, dimension(3), intent(in) :: v_i, v_j
    double precision, dimension(3), intent(in) :: omega_i, omega_j
    double precision, dimension(3), intent(out) :: v_i1, v_j1
    double precision, dimension(3), intent(out) :: omega_i1, omega_j1
    double precision, dimension(3), parameter :: z_loc = (/ 0.d0, 1.d0, 0.d0 /)
    double precision, dimension(3) :: r_c, r_ij
    double precision :: r_ij_mag_sqrd, r_ij_mag, v_ix, v_iy, v_jx, v_jy
    double precision :: omega_ix, omega_iy, omega_iz, omega_jx, omega_jy, omega_jz
    double precision :: u_iR_x, u_iR_y, u_jR_x, u_jR_y, u_iR_xy_mag, u_jR_xy_mag
    double precision :: u_ijC_x, u_ijC_z, u_ijC_xz_mag, v_ijy, v_ijy0
    double precision :: deltaP_ix, deltaP_iy, deltaP_jx, deltaP_jy, deltaP_1, deltaP_2
    double precision :: deltaV_ix, deltaV_iy, deltaV_jx, deltaV_jy
    double precision, dimension(3) :: deltaOm_i, deltaOm_j
    double precision, dimension(3) :: y_loc, x_loc
    double precision :: W = 0, deltaW
    double precision :: W_f = INF
    double precision :: W_c = INF
    double precision, dimension(3,3) :: G, G_T
    integer :: niters = 0
    r_ij = r_j - r_i
    r_ij_mag_sqrd = sum(r_ij**2)
    r_ij_mag = sqrt(r_ij_mag_sqrd)
    y_loc = r_ij / r_ij_mag
    x_loc(1) = -y_loc(3)
    x_loc(2) = 0
    x_loc(3) = y_loc(1)
    G(1,1:3) = x_loc
    G(2,1:3) = y_loc
    G(3,1:3) = z_loc
    G_T = transpose(G)
    v_ix = sum(v_i * x_loc)
    v_iy = sum(v_i * y_loc)
    v_jx = sum(v_j * x_loc)
    v_jy = sum(v_j * y_loc)
    omega_ix = sum(omega_i * x_loc)
    omega_iy = sum(omega_i * y_loc)
    omega_iz = sum(omega_i * z_loc)
    omega_jx = sum(omega_j * x_loc)
    omega_jy = sum(omega_j * y_loc)
    omega_jz = sum(omega_j * z_loc)
    u_iR_x = v_ix + R*omega_iy
    u_iR_y = v_iy - R*omega_ix
    u_jR_x = v_jx + R*omega_jy
    u_jR_y = v_jy - R*omega_jx
    u_iR_xy_mag = sqrt(u_iR_x**2 + u_iR_y**2)
    u_jR_xy_mag = sqrt(u_jR_x**2 + u_jR_y**2)
    u_ijC_x = v_ix - v_jx - R*(omega_iz + omega_jz)
    u_ijC_z =               R*(omega_ix + omega_jx)
    u_ijC_xz_mag = sqrt(u_ijC_x**2 + u_ijC_z**2)
    v_ijy = v_jy - v_iy
    v_i1(1) = v_ix
    v_i1(2) = v_iy
    v_i1(3) = 0
    v_j1(1) = v_jx
    v_j1(2) = v_jy
    v_j1(3) = 0
    omega_i1 = (/ omega_ix, omega_iy, omega_iz /)
    omega_j1 = (/ omega_jx, omega_jy, omega_jz /)
    do while (v_ijy < 0 .or. W < W_f)
       if (u_ijC_xz_mag < 1.d-16) then
          deltaP_1 = 0
          deltaP_2 = 0
          deltaP_ix = 0
          deltaP_iy = 0
          deltaP_jx = 0
          deltaP_jy = 0
       else
          deltaP_1 = -mu_b * deltaP * u_ijC_x / u_ijC_xz_mag
          if (abs(u_ijC_z) < 1.d-16) then
             deltaP_2 = 0
             deltaP_ix = 0
             deltaP_iy = 0
             deltaP_jx = 0
             deltaP_jy = 0
          else
             deltaP_2 = -mu_b * deltaP * u_ijC_z / u_ijC_xz_mag
             if (deltaP_2 > 0) then
                deltaP_ix = 0
                deltaP_iy = 0
                if (u_jR_xy_mag == 0) then
                   deltaP_jx = 0
                   deltaP_jy = 0
                else
                   deltaP_jx = -mu_s * (u_jR_x / u_jR_xy_mag) * deltaP_2
                   deltaP_jy = -mu_s * (u_jR_y / u_jR_xy_mag) * deltaP_2
                endif
             else
                deltaP_jx = 0
                deltaP_jy = 0
                if (u_iR_xy_mag == 0) then
                   deltaP_ix = 0
                   deltaP_iy = 0
                else
                   deltaP_ix = mu_s * (u_iR_x / u_iR_xy_mag) * deltaP_2
                   deltaP_iy = mu_s * (u_iR_y / u_iR_xy_mag) * deltaP_2
                end if
             end if
          end if
       endif
       deltaV_ix = ( deltaP_1 + deltaP_ix) / M
       deltaV_iy = (-deltaP   + deltaP_iy) / M
       deltaV_jx = (-deltaP_1 + deltaP_jx) / M
       deltaV_jy = ( deltaP   + deltaP_jy) / M
       !
       deltaOm_i = 5.d0/(2*M*R) * (/ (deltaP_2 + deltaP_iy), &
                                     (-deltaP_ix), &
                                     (-deltaP_1) /)
       deltaOm_j = 5.d0/(2*M*R) * (/ (deltaP_2 + deltaP_jy), &
                                     (-deltaP_jx), &
                                     (-deltaP_1) /)
       !
       v_ix = v_ix + deltaV_ix
       v_jx = v_jx + deltaV_jx
       v_iy = v_iy + deltaV_iy
       v_jy = v_jy + deltaV_jy
       !
       omega_i1 = omega_i1 + deltaOm_i
       omega_j1 = omega_j1 + deltaOm_j
       !
       u_iR_x = v_ix + R*omega_iy
       u_iR_y = v_iy - R*omega_ix
       u_jR_x = v_jx + R*omega_jy
       u_jR_y = v_jy - R*omega_jx
       u_iR_xy_mag = sqrt(u_iR_x**2 + u_iR_y**2)
       u_jR_xy_mag = sqrt(u_jR_x**2 + u_jR_y**2)
       !
       u_ijC_x = v_ix - v_jx - R*(omega_iz + omega_jz)
       u_ijC_z = R*(omega_ix + omega_jx)
       u_ijC_xz_mag = sqrt(u_ijC_x**2 + u_ijC_z**2)
       ! increment work:
       v_ijy0 = v_ijy
       v_ijy = v_jy - v_iy
       deltaW = 0.5 * deltaP * abs(v_ijy0 + v_ijy)
       W = W + deltaW
       niters = niters + 1
       if (W_c == INF .and. v_ijy > 0) then
          W_c = W
          W_f = (1 + e**2) * W_c
       end if
    end do
    v_i1 = (/ v_ix, v_iy, 0.d0 /)
    v_j1 = (/ v_jx, v_jy, 0.d0 /)
    v_i1 = MATMUL(G_T, v_i1)
    v_j1 = MATMUL(G_T, v_j1)
    omega_i1 = MATMUL(G_T, omega_i1)
    omega_j1 = MATMUL(G_T, omega_j1)
  END SUBROUTINE collide_balls
END MODULE COLLISIONS
